pdns/pdns_recursor.cc

```diff
@@ -918,6 +920,54 @@ static void protobufLogResponse(pdns::ProtoZero::RecMessage& message)
   }
 }
 
+static void protobufLogResponse(const struct dnsheader* dh,
+                                const RecursorPacketCache::OptPBData& pbData, const struct timeval& tv,
+                                bool tcp, const ComboAddress& source, const ComboAddress& destination,
+                                const EDNSSubnetOpts& ednssubnet,
+                                const boost::uuids::uuid& uniqueId, const string& requestorId, const string& deviceId,
+                                const string& deviceName) 
+{
+  pdns::ProtoZero::RecMessage pbMessage(pbData ? pbData->d_message : "", pbData ? pbData->d_response : "", 64, 10); // The extra bytes we are going to add
+  if (pbData) {
+    // We take the inmutable string from the cache and are appending a few values
+  } else {
+    pbMessage.setType(pdns::ProtoZero::Message::MessageType::DNSResponseType);
+    pbMessage.setServerIdentity(SyncRes::s_serverID);
+  }
+
+  // In response part
+  if (g_useKernelTimestamp && tv.tv_sec) {
+    pbMessage.setQueryTime(tv.tv_sec, tv.tv_usec);
+  }
+  else {
+    pbMessage.setQueryTime(g_now.tv_sec, g_now.tv_usec);
+  }
+
+  auto luaconfsLocal = g_luaconfs.getLocal(); 
```

Comment:
Perhaps we could pass the local state holder to the function? `GlobalStateHolder::getLocal` is cheap but the next dereference of `luaconfsLocal` will call `GlobalStateHolder::getState` which is not.